### PR TITLE
[Bug Fix] Fix sparse opt bug

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Fix bug in #2856 

And also add some unitests

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 1]
```diff
@@ -77,8 +77,10 @@ def step(self):
                 for i, data in emb._trace:
                     idx.append(i)
                     grad.append(data.grad.data)
-                idx = th.cat(idx, dim=0)
-                grad = th.cat(grad, dim=0)
+                idx = th.cat(idx, dim=0) if len(idx) != 0 else \
```
Comment (zheng-da) [inline python/dgl/optim/pytorch/sparse_optim.py:80]: it's better to write comments here to explain

File: tests/pytorch/test_optim.py [hunk 2]
```diff
@@ -42,5 +45,173 @@ def test_sparse_adam():
...
+    dgl_adam.zero_grad()
+    torch_adam.zero_grad()
+    dgl_loss = th.nn.functional.cross_entropy(dgl_value, labels)
+    torch_loss = th.nn.functional.cross_entropy(torch_value, labels)
+    dgl_loss.backward()
+    torch_loss.backward()
+
+    dgl_adam.step()
+    torch_adam.step()
+    assert F.allclose(dgl_emb.emb_tensor, torch_emb.weight)
```
Comment (zheng-da) [inline tests/pytorch/test_optim.py:83]: should we rename `emb_tensor` to `weight` as well?

File: tests/pytorch/test_optim.py [hunk 3]
```diff
@@ -42,5 +45,173 @@ def test_sparse_adam():
...
+
+    dgl_adam.step()
+    torch_adam.step()
+    if rank == 0:
+        after_step = dgl_emb(idx, device)
+        assert F.allclose(dgl_emb.emb_tensor, torch_emb.module.weight)
+        assert F.allclose(dgl_value, after_step) is False
+    th.distributed.barrier()
+
+def start_sparse_adam_worker_zero_step(rank, world_size, num_embs=128, emb_dim=10):
```
Comment (zheng-da) [inline tests/pytorch/test_optim.py:133]: what's the difference between this one and the one above?
also can we merge it with the one above. it seems most of the code is the same.

File: tests/pytorch/test_optim.py (hunk 2 above)
Comment (classicsong) [reply tests/pytorch/test_optim.py:83]: Good point.

**Review by zheng-da — APPROVED** [review]

**Pull request merged**